# Module 10: Descriptive Statistics and Outliers
***

## What are statistics?

Statistics deal with collecting informative data, interpreting these data, and drawing conclusions about the data. Statistics are used across multiple disciplines and fuel business and policy decisions.

## Two types of statistics
***
#### Descriptive Statistics
Descriptive statistics use numeric and graphical techniques to summarize the characteristics of a set of data and to identify patterns within the dataset. Descriptive statistics also include graphically summarizing data and presenting the information in a meaningful (easy to interpret) way. 

#### Inferential Statistics
Inferential statistics use samples of data to make generalizations about a larger set of data (or a larger population). Inferential statistics use data to make estimates, decisions, and predictions about patterns and trends identified within a set of data. Inferential statistics are used to make conclusions beyond your immediate set of data. 

## Import Libraries and Data

In [1]:
import pandas as pd
import numpy as np
import scipy.stats as stats ## new library alert! ##

## SciPy is used for scientific and mathematic computing; it provides functions for stats, 
# data manipulation, and data visualization

df = pd.read_excel("axisdata.xlsx")

df.head()

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0  Jackie    Jackson  Trainee      F            32             Y   
1    Mary  Patterson  Trainee      F            43             N   
2   Tanya      Adams  Trainee      F            28             Y   
3   Tanya  Henderson  Trainee      F            24             Y   
4  Walter   Franklin  Trainee      M            25             Y   

   Years Experience  Cars Sold  
0                 1          3  
1                 1          6  
2                 1          2  
3                 1          5  
4                 1          4

## Qualitative vs Quantitative Data
***

#### Quantitative Data
Data that are recorded on a numeric scale (i.e. age, weight, salary, number of empty seats on a plane, credit score etc). Quantitative variables are also called <b>numeric</b> variables. 

#### Qualitative Data
Data that cannot be measured on a numeric scale (i.e. eye color, type of car, blood type, race etc). These data can only be classified into one group of categories. Qualitative variables are also called <b>categorical</b> variables. Qualitative data also includes text and string data. 

In [2]:
## Which variables are qualitative and which variables are quantitative?

df.head()

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0  Jackie    Jackson  Trainee      F            32             Y   
1    Mary  Patterson  Trainee      F            43             N   
2   Tanya      Adams  Trainee      F            28             Y   
3   Tanya  Henderson  Trainee      F            24             Y   
4  Walter   Franklin  Trainee      M            25             Y   

   Years Experience  Cars Sold  
0                 1          3  
1                 1          6  
2                 1          2  
3                 1          5  
4                 1          4

## Describing Qualitative Data
***

Qualitative data is best summarized by frequency tables. 

Frequency tables summarize categorical data into a table showing each category, number of observations that fall within each category, and the relative frequency of each category. 

Relative frequency is the number of specific observations out of the total number of observations, or the percentage of the total sample that fall within a specific group. Both representations of categorical data give insight into how common specific categories are within a set of data. 

<b>Bar charts</b> are ideal for graphically representing qualitative data as they visually show the frequency of observations across various groups. 

#### GRAPHIC SHOULD READ N = 150 !!!

<img src="FreqTable.png">

In [3]:
## Frequencies 

df["Gender"].value_counts()

M    510
F    489
Name: Gender, dtype: int64

In [4]:
## Relative Frequencies 

df["Gender"].value_counts(normalize=True)

M    0.510511
F    0.489489
Name: Gender, dtype: float64

In [5]:
## Multi-variable Frequencies 

pd.crosstab(df["Gender"], df["SalesTraining"], margins=False, normalize=False)

SalesTraining    N    Y
Gender                 
F              212  277
M              201  309

# { Exercise 1 }

    1. Import the "pokemon.csv" file; name the dataset 'poke'. Preview the first 5 rows. 
    2. Determine the frequencies of unique pokemon types (variable name: Type 1) in this dataset
    3. Determine the relative frequencies of the unique pokemon types (variable name: Type 1) in this dataset
    4. Create a crosstabs table that shows the frequency of pokemon stage (variable name: Stage) and legendary status (variable name: Legendary). 

In [6]:
poke = pd.read_csv("pokemon.csv")
poke.head()

Num        Name Type 1  Type 2  Total  HP  Attack  Defense  SpAtk  SpDef  \
0    1   Bulbasaur  Grass  Poison    318  45      49       49     65     65   
1    2     Ivysaur  Grass  Poison    405  60      62       63     80     80   
2    3    Venusaur  Grass  Poison    525  80      82       83    100    100   
3    4  Charmander   Fire     NaN    309  39      52       43     60     50   
4    5  Charmeleon   Fire     NaN    405  58      64       58     80     65   

   Speed  Stage  Legendary  
0     45      1      False  
1     60      2      False  
2     80      3      False  
3     65      1      False  
4     80      2      False

In [7]:
#Determine the frequencies of unique pokemon types (variable name: Type 1) in this dataset
poke["Type 1"].value_counts()

Water       28
Normal      22
Poison      14
Grass       12
Fire        12
Bug         12
Electric     9
Rock         9
Ground       8
Psychic      8
Fighting     7
Ghost        3
Dragon       3
Fairy        2
Ice          2
Name: Type 1, dtype: int64

In [8]:
#Determine the relative frequencies of the unique pokemon types (variable name: Type 1) in this dataset
poke["Type 1"].value_counts(normalize=True)

Water       0.185430
Normal      0.145695
Poison      0.092715
Grass       0.079470
Fire        0.079470
Bug         0.079470
Electric    0.059603
Rock        0.059603
Ground      0.052980
Psychic     0.052980
Fighting    0.046358
Ghost       0.019868
Dragon      0.019868
Fairy       0.013245
Ice         0.013245
Name: Type 1, dtype: float64

In [9]:
#Create a crosstabs table that shows the frequency of pokemon stage (variable name: Stage) and legendary status (variable name: Legendary). 
pd.crosstab(poke["Stage"], poke["Legendary"], margins=True, normalize=False)

Legendary  False  True  All
Stage                      
1             75     4   79
2             56     0   56
3             16     0   16
All          147     4  151

## Describing Quantitative Data

There are many more techniques for describing quantitative data compared to qualitative. When summarizing quantitative data, the objective is to get a sense of how the data is clustered and the variation in numeric values. 

***

### Measures of Central Tendency

Measures of Central Tendency are estimates of what a "typical" value/data point is within your set of data. A measure of central tendency is a number around which an entire sample of data is spread. We can describe this central position using several statistics, including the mean (average), the median, and the mode. 

Measures of central tendency can be visualized with a frequency distribution (histogram). A frequency distribution is a description of values and how often each value occurs within your dataset. When you graph a frequency distribution a pattern/shape should give insight into which values occured the most/least. 

<img src="NormalDist.jpg">

***

#### Mean

The mean, which is the most commonly known and used measure of central tendency, is the arithmetic average of a set of numeric values. The mean is calculated by adding all the values together and dividing by the total number of values in the set of values. In specific sets of data, the mean is the central point in the frequency distribution aka the hump. 

In [10]:
## Calculate the Mean of a column 

df["Hours Worked"].mean()

33.727727727727725

#### Median 

The median is the value which divides the data into two equal parts when the data is arranged from least to greatest. The median will be the middle value if the number of values is odd, otherwise, it will be the average of the two middle values if the number of values is even. 

In [11]:
## Calculate the Median of a column 

df["Hours Worked"].median()

34.0

#### Mode 

The mode is the value that appears most frequently in a sample of data, or the value that has the highest frequency. It is possible for a set of data to have no mode, and it is possible for a set of data to have multiple modes. 

In [12]:
## Calculate the Mode of a column 

df["Hours Worked"].mode()

0    38
Name: Hours Worked, dtype: int64

# { Exercise 2 }

    1. Using the pokemon dataset ('poke'), calculate the mean attack value (Attack) in the dataset.  
    2. Determine the median defense value (Defense). Write a single sentence on what your findings mean. 
    3. Determine the mode of the speed (Speed) column. Write a single sentence on what your findings mean. 

In [13]:
poke["Attack"].mean()

72.54966887417218

In [14]:
poke["Defense"].median()
print("the median defense value is ", poke["Defense"].median())

the median defense value is  65.0


In [15]:
poke["Speed"].mode()
print("the mode of the speed column is ", poke["Speed"].mode())

the mode of the speed column is  0    70
1    90
Name: Speed, dtype: int64


### Data Distribution and Measurement Sensitivity
***

### Normally distributed data vs. non-normally distributed data

Several statistics are suited for data that follows a <b>normal distribution</b>. When visualized, normally distributed data lies in a symmetrical pattern (bell-shape) where the majority of the data surrounds the mean with decreasing amounts evenly distributed to the left and the right. 

While this is the ideal, real-world data is not always normally distributed. Non-normal distribution is possible and there are several statistics that can still be used with non-normally distributed data. It is important to determine how your data is distributed before you embark on certain statistics. 

#### Distribution and Measure of Central Tendency

The shape of the data helps to determine which measure of central tendency should be reported. Some statistics are very sensitive to non-normal distributions, while others are more resilient. Three of the most common shapes to learn are symmetric, left-skewed, and right-skewed. <b>Skew</b> is simply a measure of the asymmetry of the distribution. <b>In the below images, while not shown, the values are increasing from left to right.</b> 

#### Symmetric Data
* mean, median, and mode are all the same here
* no skewness is apparent
* the distribution is described as symmetric

<img src="Symmetrical.jpg">

***

#### Left-skewed Data
* mean < median
* long tail on the left

<img src="LeftSkew.jpg">

***

#### Right-skewed Data
* mean > median
* long tail on the right

<img src="RightSkew.jpg">

***

### Considerations for Shape

When your data is skewed in either direction, you should consider the best measure of central tendency to report. The median is best used when you have skewed data. This is because the median is less affected by extreme values. 

If your data is skewed, this could be an indication that your data has extreme high or low values that are contributing to the skewness. Later in this lesson, we will talk about how to handle extreme values. 

### Measures of Variability (or Spread)

Measures of Variability, or spread, describe the amount of dispersion in your data sample. In other words, measures of variability, describe how spread out values are. For example, in a sample of 100 students, the average test score is 75. However, not all students will have gotten a 75 - their scores will be spread out above and below the average, some will be higher and some lower. To describe the spread of our data, we can use several statistics, including range, standard deviation, percentiles, and quartiles.

***

#### Range

The range is the difference between the lowest and highest value in a set of data. For example, if the maximum value is 10 and the minimum value is 2, the range is 8. The larger the range, the larger the difference between the highest and lowest values. 

In [16]:
## Calculate the range of values in a column

hrs_range = df['Hours Worked'].max() - df['Hours Worked'].min()

print(hrs_range)

28


#### Standard Deviation (SD)

The SD is a measurement of the average distance between each data value and the mean. You can consider SD as a measure of how spread out the data is from the mean. A low SD indicates that the data points are clustered around the mean, while a high SD indicates the data points are spread out over a wider range of values. 

<img src="https://d20khd7ddkh5ls.cloudfront.net/high_low_standard_deviation.png">

In [17]:
## Calculate the Standard Deviation of values in a column

print("Mean hours worked by all employees:", df["Hours Worked"].mean())

df["Hours Worked"].std()

Mean hours worked by all employees: 33.727727727727725


8.223453795492198

#### Percentiles

Percentiles give insight into the relation of one data point to other data points within the same dataset. When the data values are arranged from lowest to highest, percentiles represent the position of a value within the list of values. For example, if a value is at the 50th percentile - 50% of the data falls below the value, and 50% fall above the value. If a value is at the 85th percentile - 85% of the data falls below the value, and 15% fall above the value. 

In [18]:
## Calculate percentiles of values in a column
## To determine the value that falls at a specific percentile
## np.percentile(data/column, percentile value)

np.percentile(df['Hours Worked'], 85)

## Determine which value falls at the 85th percentile? 
## 85% of the values in Hours Worked are below this value, 15% are above this value. 

44.0

#### Quartiles 

Quartiles are values that divide a set of data into equal quarters (when the data is ordered from lowest to highest). There are <b>three</b> quartile values that divide the data into four parts (see below). 

<img src="quar.png">

* The first quartile (Q1) is at the 25th percentile (25% below, 75% above)
* The second quartile (Q2) is at the 50th percentile (50% below, 50% above, median)
* The third quartile (Q3) is at the 75th percentile (75% below, 25% above)

<b>Quartiles can be used to calculate the Interquartile Range (IQR) - which can be used to handle outliers in your dataset.</b>

In [19]:
## Determine the three quartiles of a set of data
## A quantile is a series of values, or cut points, that divide a set of data into equal parts 
## A 'quartile' is a specical kind of quantile that divides the data into four equal parts

df['Hours Worked'].quantile(.25) 

#Q1 = .25
#Q2 = .50
#Q3 = .75

27.0

#### All of the Above 

The describe function is perfect for outputting multiple descriptive statistics at once. This will save you multiple steps if you want to look at one or several quantitative variables. 

In [20]:
## Use the describe function to view multiple statistics at the same time

df.describe()

Hours Worked  Years Experience   Cars Sold
count    999.000000        999.000000  999.000000
mean      33.727728          3.026026    3.922923
std        8.223454          1.394709    1.527000
min       20.000000          1.000000    1.000000
25%       27.000000          2.000000    3.000000
50%       34.000000          3.000000    4.000000
75%       41.000000          4.000000    5.000000
max       48.000000          5.000000    7.000000

In [21]:
## Use the describe function to isolate information for a single column

df['Hours Worked'].describe()

count    999.000000
mean      33.727728
std        8.223454
min       20.000000
25%       27.000000
50%       34.000000
75%       41.000000
max       48.000000
Name: Hours Worked, dtype: float64

# { Exercise 3 }

    1. Using the pokemon dataset ('poke'), calculate the range for the HP column. 
    2. What is the mean and the standard deviation for the special attack (SpAtk) column?
    3. For the Attack column, determine which value falls at the 80th percentile? Which value falls at the 15th percentile? For each of these values, write a single sentence describing what this means. 
    4. Calculate the second quartile (or the median) for the special defense (SpDef) column. Then, calculate the median of the same column using the method shown above. Are the values the same? 

In [22]:
poke_range = poke['HP'].max() - poke['HP'].min()

print(poke_range)

240


In [23]:
print("Mean special attack for pokemon is:", poke["SpAtk"].mean())

poke["SpAtk"].std()

Mean special attack for pokemon is: 67.13907284768212


28.53419930191353

In [24]:
print(poke['Attack'].quantile(.80))
print(poke['Attack'].quantile(.15)) 

95.0
45.0


In [25]:
#Calculate the second quartile (or the median) for the special defense (SpDef) column. 
print(poke['SpDef'].quantile(.50))

65.0


In [26]:
#calculate the median of the same column using the method shown above. 
poke['SpDef'].median()

65.0

In [27]:
#The values are the same: 65.0

## Handling and Removing Outliers

***

Outliers are data points that differ significantly from other data points. Outliers are an important feature of a dataset that should be identified and addressed because extreme values can influence the results of your statistics. 

   * In a sample of 5 students, their ages are 25, 22, 97, 23, 21 - 97 is an outlier because the value differs significantly from the remaining data.
   * Without removing outlier, the mean is: 37.6 and the median is: 23
   * When removing the outlier, the mean is: 22.75 and the median is: 22.5

Determining which values are extreme can be challenging when you have a larger dataset. There are several methods that can be used to strategically determine which values are outliers, we will cover two separate methods below. 

In [28]:
df = pd.read_csv("SalaryData.csv")
df.head()

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
0  Richard              5.1    66029   41  Female             0  Central   
1     Nick              5.9  8136300   39  Female             0  Central   
2   Morgan              9.5   116969   50  Female             1  Central   
3    Susan              3.2    64445   27    Male             0  Central   
4  Matthew              8.2   113812   47  Female             1  Central   

  CompanyCar  Shift  PerformanceReviewScore  
0          Y      2                    98.3  
1          Y      3                   105.7  
2          N      1                   145.0  
3          Y      3                    75.1  
4          N      2                   131.1

## Method 1 : Using z-scores to detect and remove outliers

In statistics, the <font color=salmon><b>Empirical Rule</b></font> estimates how data, following a normal distribution, will spread around the mean. According to this rule, basically all data values will fall within 3 SD's above or below the mean. The Empirical Rule is also known as the <font color=salmon><b>68-95-99.7 Rule</b></font> because according to this rule, your data will typically fall into the following pattern:

* 68% of the data falls within one SD of the mean

* 95% of the data falls within two SD's of the mean

* 99.7% of the data falls within three SD's of the mean

<img src="SD Bell Curve.jpeg">

#### Empirical Rule Example:

The average weight of an American man is 175 lbs with a SD of 10 lbs. According to the Empirical Rule, we can assume: 

* 68% of American men are between 165 and 185 lbs (or +/- 10 lbs from the mean, or 1 SD)
* 95% of American men are between 155 and 195 lbs (or +/- 20 lbs from the mean, or 2 SD's)
* 99.7% of American men are between 145 and 205 lbs (or +/- 30 lbs from the mean, or 3 SD's)
* Typically, data values that fall outside of this range are considered outliers
_______________________ 
A <font color=salmon><b>z-score</b></font>, also known as a standard score, measures the distance of a value from the mean. In other words, the z-score tells you how many SD's a given value is from the mean. Values that fall above the mean will have a positive (+) z-score, while values that fall below the mean will have a negative (-) z-score. Values that are far from the mean will be considered outliers. You can define a value as an outlier if the z-score greater than +/- 3. 
_______________________ 

#### Steps for Detecting and Removing Outliers with z-scores
* Import the scipy library to make use of the z-score function
* Calculate the z-score for each value within a specific column in your dataset
* Transform the z-score to the absolute value of the z-score (number without the +/- sign; this will make our code simpler)
* Add the z-scores to your dataset as a new column
* Remove all rows that have a z-score greater than 3

In [29]:
## Create a copy of your dataset to filter outliers
dfz = df.copy()

## Check original shape of the dataset
print(dfz.shape)

(30, 10)


In [30]:
## Create a new column to contain the z-scores for the employee salary
## Set the new column equal to the absolute value of the z-scores
## To calculate z-score -- stats.zscore(data/column name)
## To determine absolute value of z-score -- np.abs(data/function/options)

dfz["zscore_salary"] = np.abs(stats.zscore(dfz["Salary"]))

## Preview new column; optional
dfz.head()

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
0  Richard              5.1    66029   41  Female             0  Central   
1     Nick              5.9  8136300   39  Female             0  Central   
2   Morgan              9.5   116969   50  Female             1  Central   
3    Susan              3.2    64445   27    Male             0  Central   
4  Matthew              8.2   113812   47  Female             1  Central   

  CompanyCar  Shift  PerformanceReviewScore  zscore_salary  
0          Y      2                    98.3       0.269844  
1          Y      3                   105.7       3.343557  
2          N      1                   145.0       0.247036  
3          Y      3                    75.1       0.270553  
4          N      2                   131.1       0.248449

In [31]:
## Determine the index locations for the rows with zscores that are greater than "3"
z_outliers = dfz.loc[dfz["zscore_salary"] > 3].index

## Preview list of index values
print(z_outliers)

Int64Index([1, 28], dtype='int64')


In [32]:
## What information can we find at these index locations?

dfz.iloc[[1, 28]]

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
1      Nick              5.9  8136300   39  Female             0  Central   
28     Bill              7.1  9827300   36  Female             0     East   

   CompanyCar  Shift  PerformanceReviewScore  zscore_salary  
1           Y      3                   105.7       3.343557  
28          Y      3                   116.8       4.100689

In [33]:
## Drop rows with above index values
dfz = dfz.drop(z_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfz.shape)

(28, 11)


## Method 2 : Using the Interquartile Range (IQR) to detect and remove outliers

The <font color=salmon><b>Interquartile Range (IQR)</b></font> is the difference between the first and third quartile in a set of data. The IQR can be considered where the bulk of the data values rest. IQR is calculated by the following equation: <B>IQR = Q3 - Q1</B>
<img src="IQR.png">
_____________________

The IQR can be used to create "fences" or cut-off values that determine which values fall outside of the acceptable data range. For example, we can calculate a lower-limit -- all values falling below this limit will be considered an outlier. The <b>1.5 Rule</b> is a commonly used method for determining the upper and lower cut-off limits. According to this rule, a value is considered an outlier if it falls:

* ... below Q1 - (1.5 x IQR)

* ... above Q3 + (1.5 x IQR)
_____________________

#### 1.5 Rule Example:

A professor is tracking the number of absences for each of their students. They have compiled a list of the number of absences: 

    1, 3, [4], 6, 7, [7], 8, 8, [9], 22, 37 

In this list of data, we can find the following values:

* Q1 = 4
* Q2 (median) = 7
* Q3 = 9

With these values, you can easily calculate the IQR:

* <B>IQR</B> = Q3(9) - Q1(4)
* <B>IQR</B> = 5

Once you have the IQR, you can calculate the upper and lower limits for detecting outliers:

* Upper Limit = Q3(9) + [1.5 * IQR(5)] = <b> 16.5 </b>
* Lower Limit = Q1(4) - [1.5 * IQR(5)] = <b> -3.5 </b>

Values that fall above the upper limit are outliers; values that fall below the lower limit are outliers
_______________________ 

#### Steps for Detecting and Removing Outliers with IQR

* Calculate quartiles
* Calculate IQR
* Determine upper and lower limits
* Drop values less than the lower and/or greater than the upper

In [34]:
## Create a copy of original dataset
dfq = df.copy()

## Check original shape of the dataset
dfq.shape

(30, 10)

In [35]:
## Calculate quartiles
q1 = dfq["Salary"].quantile(.25)
q3 = dfq["Salary"].quantile(.75)

print("Q1:", q1)
print("Q3:", q3)

Q1: 48265.0
Q3: 104512.0


In [36]:
## Calculate the IQR
iqr = q3 - q1

print("IQR:", iqr)

IQR: 56247.0


In [37]:
## Determine outlier fences 
top = q3 + (iqr * 1.5)
bottom = q1 - (iqr * 1.5)


print("Upper Limit:", top)
print("Lower Limit:", bottom)

Upper Limit: 188882.5
Lower Limit: -36105.5


In [38]:
## Determine the index locations for rows that fall outside of outlier fences

iqr_outliers = dfq.loc[(dfq['Salary'] > top) | (dfq['Salary'] < bottom)].index

print("INDEX VALUES:", iqr_outliers)

INDEX VALUES: Int64Index([1, 12, 28], dtype='int64')


In [39]:
## what values can we find at these index locations?

dfq.iloc[[1, 12, 28]]

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
1      Nick              5.9  8136300   39  Female             0  Central   
12   Thomas              1.5   377310   46    Male             1  Central   
28     Bill              7.1  9827300   36  Female             0     East   

   CompanyCar  Shift  PerformanceReviewScore  
1           Y      3                   105.7  
12          Y      1                    63.8  
28          Y      3                   116.8

In [40]:
## Drop rows with above index values
dfq = dfq.drop(iqr_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfq.shape)

(27, 10)


# { Module 10 Homework }

1. Import the "babies.xlsx" dataset. See below for information on the columns:

    * bwt - birth weight of newborn baby
    * gestation	- gestation length (weeks)
    * parity - previously pregnant (0 = no; 1 = yes)
    * age - age of mother
    * height - height of mother (inches)	
    * weight - weight of mother (pounds)
    * smoke - smoking status of mother (0 = nonsmoker; 1 = smoker)

In [105]:
babies = pd.read_excel("babies.xlsx")

# a = babies
a = babies.copy()
a.head()

bwt  gestation  parity   age  height  weight  smoke
0  120      284.0       0  27.0    62.0   100.0    0.0
1  113      282.0       0  33.0    64.0   135.0    0.0
2  128      279.0       0  28.0    64.0    77.0    1.0
3  123        NaN       0  36.0    69.0   190.0    0.0
4  108      282.0       0  23.0    67.0   125.0    1.0

2. Preview the first few rows of the dataset. Complete the following checks for your dataset:

    * output the summary information for the dataset, what types of variables make up this dataset?
    * is there any missing data in this dataset?

In [106]:
a.info() #types of variables: integers, floats; parity, smoke = qualitative

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1236 entries, 0 to 1235
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   bwt        1236 non-null   int64  
 1   gestation  1223 non-null   float64
 2   parity     1236 non-null   int64  
 3   age        1234 non-null   float64
 4   height     1214 non-null   float64
 5   weight     1200 non-null   float64
 6   smoke      1226 non-null   float64
dtypes: float64(5), int64(2)
memory usage: 67.7 KB


In [107]:
#rows and columns
a.shape 

(1236, 7)

In [108]:
#missing data in each column
a.isnull().sum() 

bwt           0
gestation    13
parity        0
age           2
height       22
weight       36
smoke        10
dtype: int64

In [109]:
a.isnull()

bwt  gestation  parity    age  height  weight  smoke
0     False      False   False  False   False   False  False
1     False      False   False  False   False   False  False
2     False      False   False  False   False   False  False
3     False       True   False  False   False   False  False
4     False      False   False  False   False   False  False
...     ...        ...     ...    ...     ...     ...    ...
1231  False      False   False  False   False   False  False
1232  False      False   False  False   False   False  False
1233  False      False   False  False   False   False  False
1234  False      False   False  False   False   False  False
1235  False      False   False  False   False   False  False

[1236 rows x 7 columns]

In [110]:
a.isnull().any(axis=0)

bwt          False
gestation     True
parity       False
age           True
height        True
weight        True
smoke         True
dtype: bool

In [111]:
#missing data in 62 rows

a.loc[a.isnull().any(axis = 1)]

bwt  gestation  parity   age  height  weight  smoke
3     123        NaN       0  36.0    69.0   190.0    0.0
39    128      282.0       0  31.0    65.0     NaN    0.0
42    138      302.0       0  26.0     NaN     NaN    1.0
85    129      274.0       0  29.0    71.0     NaN    1.0
89    114        NaN       0  24.0    67.0   113.0    1.0
...   ...        ...     ...   ...     ...     ...    ...
1044  117      323.0       0  26.0    62.0     NaN    0.0
1177  146      305.0       0  23.0     NaN     NaN    0.0
1190  104      275.0       0  24.0     NaN     NaN    0.0
1192  124        NaN       1  39.0    65.0   228.0    0.0
1215  152      267.0       0  28.0     NaN   119.0    1.0

[62 rows x 7 columns]

3. Handle the missing data in the dataset -- there isn't much, so we can drop all the rows that have at least one missing value. How many rows were dropped from the dataset?

In [112]:
a.dropna(inplace=True)
a

bwt  gestation  parity   age  height  weight  smoke
0     120      284.0       0  27.0    62.0   100.0    0.0
1     113      282.0       0  33.0    64.0   135.0    0.0
2     128      279.0       0  28.0    64.0    77.0    1.0
4     108      282.0       0  23.0    67.0   125.0    1.0
5     136      286.0       0  25.0    62.0    93.0    0.0
...   ...        ...     ...   ...     ...     ...    ...
1231  113      275.0       1  27.0    60.0   100.0    0.0
1232  128      265.0       0  24.0    67.0   120.0    0.0
1233  130      291.0       0  30.0    65.0    67.0    1.0
1234  125      281.0       1  21.0    65.0   110.0    0.0
1235  117      297.0       0  38.0    65.0   129.0    0.0

[1174 rows x 7 columns]

In [113]:
a.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1174 entries, 0 to 1235
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   bwt        1174 non-null   int64  
 1   gestation  1174 non-null   float64
 2   parity     1174 non-null   int64  
 3   age        1174 non-null   float64
 4   height     1174 non-null   float64
 5   weight     1174 non-null   float64
 6   smoke      1174 non-null   float64
dtypes: float64(5), int64(2)
memory usage: 73.4 KB


In [114]:
#62 rows with missing data were dropped

1236 - 1174

62

4. There are two qualitative variables in this dataset. What are they? How do you know they are qualitative variables?

In [115]:
#parity and smoke are qualitative variables, because answers are no (0) or yes (1)


5. Replace the values in the qualitative variables with meaningful labels that describe the different groups. Use any method you've learned in previous modules to complete this task. 

In [116]:
a["parity"].replace(0, "No", inplace=True)
a

bwt  gestation parity   age  height  weight  smoke
0     120      284.0     No  27.0    62.0   100.0    0.0
1     113      282.0     No  33.0    64.0   135.0    0.0
2     128      279.0     No  28.0    64.0    77.0    1.0
4     108      282.0     No  23.0    67.0   125.0    1.0
5     136      286.0     No  25.0    62.0    93.0    0.0
...   ...        ...    ...   ...     ...     ...    ...
1231  113      275.0      1  27.0    60.0   100.0    0.0
1232  128      265.0     No  24.0    67.0   120.0    0.0
1233  130      291.0     No  30.0    65.0    67.0    1.0
1234  125      281.0      1  21.0    65.0   110.0    0.0
1235  117      297.0     No  38.0    65.0   129.0    0.0

[1174 rows x 7 columns]

In [117]:
a["parity"].replace(1, "Yes", inplace=True)
a

bwt  gestation parity   age  height  weight  smoke
0     120      284.0     No  27.0    62.0   100.0    0.0
1     113      282.0     No  33.0    64.0   135.0    0.0
2     128      279.0     No  28.0    64.0    77.0    1.0
4     108      282.0     No  23.0    67.0   125.0    1.0
5     136      286.0     No  25.0    62.0    93.0    0.0
...   ...        ...    ...   ...     ...     ...    ...
1231  113      275.0    Yes  27.0    60.0   100.0    0.0
1232  128      265.0     No  24.0    67.0   120.0    0.0
1233  130      291.0     No  30.0    65.0    67.0    1.0
1234  125      281.0    Yes  21.0    65.0   110.0    0.0
1235  117      297.0     No  38.0    65.0   129.0    0.0

[1174 rows x 7 columns]

In [118]:
#smoke - smoking status of mother (0 = nonsmoker; 1 = smoker)
a['smoke'].replace(0.0, 'nonsmoker', inplace=True)
a

bwt  gestation parity   age  height  weight      smoke
0     120      284.0     No  27.0    62.0   100.0  nonsmoker
1     113      282.0     No  33.0    64.0   135.0  nonsmoker
2     128      279.0     No  28.0    64.0    77.0        1.0
4     108      282.0     No  23.0    67.0   125.0        1.0
5     136      286.0     No  25.0    62.0    93.0  nonsmoker
...   ...        ...    ...   ...     ...     ...        ...
1231  113      275.0    Yes  27.0    60.0   100.0  nonsmoker
1232  128      265.0     No  24.0    67.0   120.0  nonsmoker
1233  130      291.0     No  30.0    65.0    67.0        1.0
1234  125      281.0    Yes  21.0    65.0   110.0  nonsmoker
1235  117      297.0     No  38.0    65.0   129.0  nonsmoker

[1174 rows x 7 columns]

In [119]:
a['smoke'].replace(1.0, 'smoker', inplace=True)
a

bwt  gestation parity   age  height  weight      smoke
0     120      284.0     No  27.0    62.0   100.0  nonsmoker
1     113      282.0     No  33.0    64.0   135.0  nonsmoker
2     128      279.0     No  28.0    64.0    77.0     smoker
4     108      282.0     No  23.0    67.0   125.0     smoker
5     136      286.0     No  25.0    62.0    93.0  nonsmoker
...   ...        ...    ...   ...     ...     ...        ...
1231  113      275.0    Yes  27.0    60.0   100.0  nonsmoker
1232  128      265.0     No  24.0    67.0   120.0  nonsmoker
1233  130      291.0     No  30.0    65.0    67.0     smoker
1234  125      281.0    Yes  21.0    65.0   110.0  nonsmoker
1235  117      297.0     No  38.0    65.0   129.0  nonsmoker

[1174 rows x 7 columns]

6. Before we move forward with any statistics, let's identify and remove any outliers from the dataset. Using the IQR method, search for outliers in the 5 numeric variables. This will take some time and organization, be careful with your code! Make sure you keep track of how many rows/outliers are removed. 

In [120]:
a_copy = a.copy()

a_copy.shape

(1174, 7)

In [121]:
#Calculate quartiles
#all 5 numeric columns
print(a_copy.quantile(.25))
print(a_copy.quantile(.75))


bwt          108.0
gestation    272.0
age           23.0
height        62.0
weight       114.0
Name: 0.25, dtype: float64
bwt          131.00
gestation    288.00
age           31.00
height        66.00
weight       139.75
Name: 0.75, dtype: float64


In [122]:
#Calculate quartiles for the "age" column
q1_age = a_copy["age"].quantile(.25)
q3_age = a_copy["age"].quantile(.75)

q1_bwt = a_copy["bwt"].quantile(.25)
q3_bwt = a_copy["bwt"].quantile(.75)

q1_gestation = a_copy["gestation"].quantile(.25)
q3_gestation = a_copy["gestation"].quantile(.75)

q1_height = a_copy["height"].quantile(.25)
q3_height = a_copy["height"].quantile(.75)

q1_weight = a_copy["weight"].quantile(.25)
q3_weight = a_copy["weight"].quantile(.75)


print("Q1 Age:", q1_age)
print("Q3 Age:", q3_age)

Q1 Age: 23.0
Q3 Age: 31.0


In [123]:
#Calculate IQR (unuseful for formula)
iqr_age = q3_age - q1_age
print("IQR Age ", iqr_age)
iqr_bwt = 131.00 - 108.0
print("IQR bwt ", iqr_bwt)
iqr_gestation = 288.00 - 272.0
print("IQR gestation ", iqr_gestation)
iqr_height = 66.00 - 62.0
print("IQR height ", iqr_height)
iqr_weight = 139.75 - 114.0
print("IQR weight ", iqr_weight)

IQR Age  8.0
IQR bwt  23.0
IQR gestation  16.0
IQR height  4.0
IQR weight  25.75


In [124]:
#Calculate IQR for formula
iqr_age = q3_age - q1_age
print("IQR Age ", iqr_age)

iqr_bwt = q3_bwt - q1_bwt
print("IQR bwt ", iqr_bwt)

iqr_gestation = q3_gestation - q1_gestation
print("IQR gestation ", iqr_gestation)

iqr_height = q3_height - q1_height
print("IQR height ", iqr_height)

iqr_weight = q3_weight - q1_weight
print("IQR weight ", iqr_weight)

IQR Age  8.0
IQR bwt  23.0
IQR gestation  16.0
IQR height  4.0
IQR weight  25.75


In [125]:
#Determine upper and lower limits
upper_limit_age = q3_age + (iqr_age * 1.5)
lower_limit_age = q1_age - (iqr_age * 1.5)

print("Upper Limit Age:", upper_limit_age)
print("Lower Limit Age:", lower_limit_age)

upper_limit_bwt = q3_bwt + (iqr_bwt * 1.5)
lower_limit_bwt = q1_bwt - (iqr_bwt * 1.5)

upper_limit_gestation = q3_gestation + (iqr_gestation * 1.5)
lower_limit_gestation = q1_gestation - (iqr_gestation * 1.5)

upper_limit_height = q3_height + (iqr_height * 1.5)
lower_limit_height = q1_height - (iqr_height * 1.5)

upper_limit_weight = q3_weight + (iqr_weight * 1.5)
lower_limit_weight = q1_weight - (iqr_weight * 1.5)


Upper Limit Age: 43.0
Lower Limit Age: 11.0


In [126]:
#index location

#age
iqr_outliers_age = a_copy.loc[(a_copy['age'] > upper_limit_age) | (a_copy['age'] < lower_limit_age)].index

print("INDEX VALUES age:", iqr_outliers_age)

#bwt  
iqr_outliers_bwt = a_copy.loc[(a_copy['bwt'] > upper_limit_bwt) | (a_copy['bwt'] < lower_limit_bwt)].index

print("INDEX VALUES bwt:", iqr_outliers_bwt)


# gestation  

iqr_outliers_gestation = a_copy.loc[(a_copy['gestation'] > upper_limit_gestation) | (a_copy['gestation'] < lower_limit_gestation)].index

print("INDEX VALUES gestation:", iqr_outliers_gestation)

# height  

iqr_outliers_height = a_copy.loc[(a_copy['height'] > upper_limit_height) | (a_copy['height'] < lower_limit_height)].index

print("INDEX VALUES height:", iqr_outliers_height)

# weight  

iqr_outliers_weight = a_copy.loc[(a_copy['weight'] > upper_limit_weight) | (a_copy['weight'] < lower_limit_weight)].index

print("INDEX VALUES weight:", iqr_outliers_weight)

INDEX VALUES age: Int64Index([633, 1070], dtype='int64')
INDEX VALUES bwt: Int64Index([ 239,  309,  361,  462,  500,  529,  556,  594,  632,  709,  738,
             747,  829,  904,  912,  978, 1021, 1035, 1063, 1065, 1099, 1139,
            1148, 1169],
           dtype='int64')
INDEX VALUES gestation: Int64Index([   6,    7,   10,   59,   63,   66,  119,  129,  155,  188,  192,
             198,  210,  215,  217,  234,  240,  253,  260,  279,  345,  361,
             373,  394,  440,  460,  462,  484,  500,  511,  523,  630,  685,
             699,  710,  726,  746,  761,  769,  778,  784,  828,  829,  830,
             833,  869,  912,  927,  952,  969,  978, 1002, 1020, 1026, 1035,
            1065, 1074, 1134, 1139, 1140, 1142, 1146, 1152, 1172, 1178, 1199,
            1206, 1207, 1217, 1219, 1226],
           dtype='int64')
INDEX VALUES height: Int64Index([434, 1208], dtype='int64')
INDEX VALUES weight: Int64Index([  14,   23,   88,  117,  125,  149,  162,  181,  183,  222,  240

In [127]:
a_copy.shape

(1174, 7)

In [134]:
iqr_outliers_age, iqr_outliers_bwt, iqr_outliers_gestation

(Int64Index([633, 1070], dtype='int64'),
 Int64Index([ 239,  309,  361,  462,  500,  529,  556,  594,  632,  709,  738,
              747,  829,  904,  912,  978, 1021, 1035, 1063, 1065, 1099, 1139,
             1148, 1169],
            dtype='int64'),
 Int64Index([   6,    7,   10,   59,   63,   66,  119,  129,  155,  188,  192,
              198,  210,  215,  217,  234,  240,  253,  260,  279,  345,  361,
              373,  394,  440,  460,  462,  484,  500,  511,  523,  630,  685,
              699,  710,  726,  746,  761,  769,  778,  784,  828,  829,  830,
              833,  869,  912,  927,  952,  969,  978, 1002, 1020, 1026, 1035,
             1065, 1074, 1134, 1139, 1140, 1142, 1146, 1152, 1172, 1178, 1199,
             1206, 1207, 1217, 1219, 1226],
            dtype='int64'))

In [139]:
set(iqr_outliers_bwt) & set(iqr_outliers_gestation)

{361, 462, 500, 829, 912, 978, 1035, 1065, 1139}

In [141]:
a_drop = a_copy.copy()

print(a_drop.shape)

#Drop values less than the lower and/or greater than the upper

##errors{‘ignore’, ‘raise’}, default ‘raise’

a_drop = a_drop.drop(iqr_outliers_age)
print(a_drop.shape)

a_drop = a_copy.drop(iqr_outliers_bwt)
print(a_drop.shape)

a_drop = a_drop.drop(iqr_outliers_gestation, errors='ignore')
print(a_drop.shape)

a_drop = a_drop.drop(iqr_outliers_height, errors='ignore')
print(a_drop.shape)

a_drop = a_drop.drop(iqr_outliers_weight, errors='ignore')
print(a_drop.shape)



(1174, 7)
(1172, 7)
(1150, 7)
(1088, 7)
(1086, 7)
(1050, 7)


7. Describe the characteristics of your qualitative variables by doing the following:

    * Determine the frequencies of each group within each categorical variable. Which groups have the highest frequency?
    * Determine the relative frequencies of each group within each categorical variable. Which groups have the highest relative frequencies?
    * Create a crosstab table using both the categorical variables in your dataset. How many mothers are smokers who have been pregnant before?

In [73]:
#Determine the frequencies of each group within each categorical variable. 
#Which groups have the highest frequency? ##noparity = 847
a_copy["parity"].value_counts()


No     847
Yes    303
Name: parity, dtype: int64

In [72]:
a_copy['smoke'].value_counts() ## nonsmoker = 699

nonsmoker    699
smoker       451
Name: smoke, dtype: int64

In [74]:
#Determine the relative frequencies of each group within each categorical variable. 
#Which groups have the highest relative frequencies?
a_copy["parity"].value_counts(normalize=True)

No     0.736522
Yes    0.263478
Name: parity, dtype: float64

In [75]:
a_copy["smoke"].value_counts(normalize=True)

nonsmoker    0.607826
smoker       0.392174
Name: smoke, dtype: float64

In [143]:
#Create a crosstab table using both the categorical variables in your dataset. 
#How many mothers are smokers who have been pregnant before?

### 118 mothers are smokers who have been pregnant before

pd.crosstab(a_copy["smoke"], a_copy["parity"], margins=True, normalize=False)

parity      No  Yes   All
smoke                    
nonsmoker  525  190   715
smoker     341  118   459
All        866  308  1174

8. Describe the Measures of Central Tendency (mean, median, mode) of the numeric variables in the dataset. Calculate the mean, median, and mode for all 5 numeric variables. For each median and mode, describe the following:

    * What does the median value tell you about the data?
    * What does the mode value tell you about the data?

In [146]:
#mean

print(a_copy["age"].mean())
print(a_copy["bwt"].mean())
print(a_copy["gestation"].mean())
print(a_copy["height"].mean())
print(a_copy["weight"].mean())

27.228279386712096
119.46252129471891
279.1013628620102
64.04940374787053
131.32623509369677


In [147]:
#median
print(a_copy['age'].median())
print(a_copy['bwt'].median())
print(a_copy['gestation'].median())
print(a_copy['height'].median())
print(a_copy['weight'].median())

26.0
120.0
280.0
64.0
125.0


In [148]:
#mode
print(a_copy['age'].mode())
print(a_copy['bwt'].mode())
print(a_copy['gestation'].mode())
print(a_copy['height'].mode())
print(a_copy['weight'].mode())

0    23.0
Name: age, dtype: float64
0    115
Name: bwt, dtype: int64
0    282.0
Name: gestation, dtype: float64
0    64.0
Name: height, dtype: float64
0    130.0
Name: weight, dtype: float64


In [ ]:
#Symmetric Data¶
#mean, median, and mode are all the same here

#Left-skewed Data¶
#mean < median
#long tail on the left

#Right-skewed Data¶
#mean > median
#long tail on the right

#age: 27.2, 26.0 23.0 - Right-skewed Data
#bwt: 119.46 120.0 115 - Left-skewed Data
#gestation: 279.1, 280.0, 282.0 - Right-skewed Data
#height: 64.049, 64.0, 64.0 - Symmetric Data
#weight: 131.326,125.0,130.0 - Right-skewed Data

9. Describe the Measures of Variability of the numeric variables in the dataset. Calculate the range, standard deviation, and 85th percentile for all 5 numeric variables. Answer the following questions:

    * Which column has the largest range? Which has the smallest? What does this mean?
    * Which column has the largest standard deviation? Which has the smallest? What does this tell you about those variables?

In [155]:
#range, standard deviation, and 85th percentile 
##age
range_age = a_copy['age'].max() - a_copy['age'].min()

print("range", range_age)

standard_deviation= a_copy["age"].std()

print("standard_deviation", standard_deviation)

percentile_age = a_copy['age'].quantile(.85)
print("percentile_age", percentile_age)

range 30.0
standard_deviation 5.817838716855792
percentile_age 34.0


In [156]:
#bwt
range_bwt = a_copy['bwt'].max() - a_copy['bwt'].min()

print("range", range_bwt)

standard_deviation= a_copy["bwt"].std()

print("standard_deviation", standard_deviation)

percentile_bwt = a_copy['bwt'].quantile(.85)
print("percentile_bwt", percentile_bwt)

range 121
standard_deviation 18.32867144332098
percentile_bwt 138.0


In [159]:
#gestation
range_gestation = a_copy['gestation'].max() - a_copy['gestation'].min()

print("range", range_gestation)

standard_deviation= a_copy["gestation"].std()

print("standard deviation", standard_deviation)

percentile_gestation = a_copy['gestation'].quantile(.85)
print("percentile", percentile_gestation)


range 205.0
standard deviation 16.010305127129392
percentile 292.0


In [160]:
#height
range_gestation = a_copy['height'].max() - a_copy['height'].min()

print("range", range_gestation)

standard_height= a_copy["height"].std()

print("standard deviation", standard_deviation)

percentile_height = a_copy['height'].quantile(.85)
print("percentile", percentile_height)

range 19.0
standard deviation 16.010305127129392
percentile 67.0


In [161]:
#weight
range_weight = a_copy['weight'].max() - a_copy['weight'].min()

print("range", range_gestation)

standard_weight= a_copy["weight"].std()

print("standard deviation", standard_deviation)

percentile_weight = a_copy['weight'].quantile(.85)
print("percentile", percentile_weight)

range 19.0
standard deviation 16.010305127129392
percentile 148.0


In [ ]:
## the largest range range has the column "gestation" = 205.0; 
## the smallest - two columns "weight" and "height" - 19.0. 
##The range is the difference between the lowest and highest value in babies data_set. 
##The larger the range, the larger the difference between the longest and shortest periods of pregnancy (weeks)

## the largest standard deviation has the "bwt" column - 18.32867. The bwt points are spread out over a wider range of values.  
##the smallest - "age" column - 5.81. Age points are clustered around the mean

#What does this tell you about those variables?
##The SD is a measurement of the average distance between each data value and the mean. SD is a measure of how spread out the data is from the mean. A low SD indicates that the age points are clustered around the mean. High SD indicates the data points are spread out over a wider range of values.